### PR TITLE
Prevent runtime from using messages with version 127

### PR DESCRIPTION
**Pull request by askibin**
#### Problem
To prevent social attacks by which the signer is tricked into signing a transaction, the runtime **MUST NOT** accept signed off-chain messages as transactions under any circumstances. The first byte of the [signing domain specifier](#signing-domain-specifier) is chosen such that it corresponds to a value (`0xff`) which is implicitly illegal as the first byte in a transaction `MessageHeader` today. The property is implicit because the top bit in the first byte of a `MessageHeader` being set signals a versioned transaction, but only a value of zero is supported at this time. The runtime needs to be modified to reserve 127 as an illegal version number, making this property explicit.

#### Summary of Changes
Deserialization of VersionedMessage

Fixes #

### CODE REVIEW

**Review by CriesofCarrots — APPROVED** [review]
Looks reasonable to me, but would be good to get @t-nelson to sanity check as well.

File: sdk/program/src/message/versions/mod.rs [hunk 1]
```diff
@@ -266,18 +266,26 @@ impl<'de> Deserialize<'de> for VersionedMessage {
...
+                                        de::Error::invalid_length(1, &self)
+                                    },
+                                )?))
+                            }
+                            127 => {
+                                // 0xff is used as the first byte of the off-chain messages
+                                // which corresponds to version 127 of the versioned messages.
+                                // This explicit check is added to prevent the usage of version 127
+                                // in the runtime as a valid transaction.
+                                Err(de::Error::custom("off-chain messages are not accepted"))
```
Comment (t-nelson) [inline sdk/program/src/message/versions/mod.rs:283]:
```suggestion
                                Err(de::Error::invalid_value(
                                    de::Unexpected::Unsigned(version as u64),
                                    &"illegal version! [0] is reserved for discriminating offchain signed messages from transactions",
                                ))
```

**Review by t-nelson — APPROVED** [review]
r+ suggested error string elaboration. thanks!

File: sdk/program/src/message/versions/mod.rs (hunk 1 above)
Comment (askibin) [reply sdk/program/src/message/versions/mod.rs:283]: I thought of using de::Error::invalid_value, but the second argument to it is the expected value. And the error gets printed like this: `Error::custom(format_args!("invalid value: {}, expected {}", unexp, exp))`. So, if you would prefer invalid_value it should be something like this:
```
Err(de::Error::invalid_value(
      de::Unexpected::Unsigned(version as u64), 
      &"a valid message version"))
```

Comment (t-nelson) [reply sdk/program/src/message/versions/mod.rs:283]: ah that's lame.  toss "transaction" in front of message on your suggestion and :shipit:

**Review by CriesofCarrots — APPROVED** [review]

**Pull request merged**